designs/odcr.md

```diff
@@ -0,0 +1,213 @@
+      #   * unlimited - The Capacity Reservation remains active until you explicitly
+      #   cancel it.
+      #   * limited - The Capacity Reservation expires automatically at a specified
+      #   date and time.
+      endDateType: String
+      # ID of the Capacity Reservation
+      id: String
+      # Indicates the type of instance launches that the Capacity Reservation accepts. The options include:
+      #   - open:
+      #       The Capacity Reservation accepts all instances that have
+      #       matching attributes (instance type, platform, and Availability
+      #       Zone). Instances that have matching attributes launch into the
+      #       Capacity Reservation automatically without specifying any
+      #       additional parameters.
+      #   - targeted:
+      #       The Capacity Reservation only accepts instances that
+      #       have matching attributes (instance type, platform, and
+      #       Availability Zone), and explicitly target the Capacity
+      #       Reservation. This ensures that only permitted instances can use
+      #       the reserved capacity.
+      instanceMatchCriteria: String
+      # Instance Platform of the Capacity Reservation
+      instancePlatform: String
+      # Instance Type of the Capacity Reservation
+      instanceType: String
+      # Owner Id of the Capacity Reservation
+      ownerId: String
+      # The date and time at which the Capacity Reservation was started.
+      startDate: String
+      # Total Instance Count of the Capacity Reservation
+      totalInstanceCount: Integer
+```
+
+This follows closely (does not implement all fields) how EC2 [DescribeCapacityReservations](https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_DescribeCapacityReservations.html) can filter.
+
+Karpenter will perform validation against the spec to ensure there isn't any violation prior to creating the LaunchTemplates.
+
+### Adding new label `karpenter.k8s.aws/capacity-reservation-id` nodeClaim/node
+
+In order for Karpenter (and admins) to understand if a NodeClaim/Node is part of an ODCR, Karpenter is adding the annotation `karpenter.k8s.aws/capacity-reservation-id` containing the Capacity Reservation Id (for example `cr-12345678`)
+
+### Launching Nodes into Capacity Reservation
+
+#### All Launch Templates are associated with the specified Capacity Reservation
+
+EC2NodeClass currently supports automatically generating Launch Templates based on instance types and their AMI requirements. We are implementing a prioritization within Karpenter for Capacity Reservation (as their price will be set to 0 making it be selected over on-demand and spot). We add a check to ensure before calling CreateFleet API existing available Capacity Reservations are being used. In some raise condition scenarios CreateFleet when creating instances against a single targeted Capacity Reservation will fail if the Reservation is fully utilized, resulting in an `ReservationCapacityExceeded` error.
```

Comment:
Saying that the price is set to `0` is somewhat inaccurate IIRC. We are going to make it "basically 0" but we should document that we are going to be doing something like "dividing all values by 1,000,000" or something so that we have miniscule numbers, but we still have some logical ordering between OD instance types if we could choose between multiple of them (this is especially important for consolidation -- where we wouldn't consolidate from one ODCR to another unless we hae some relative ordering of the pricing)